In [23]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [24]:
movies = pd.read_csv('D:\\Datasets\\Recommender_system\\movies.csv')

In [25]:
movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [26]:
movies['Year'] = movies['title'].str.extract(pat = ("(\d\d\d\d)"), expand=False)
movies['title'] = movies.title.str.replace(pat = "(\(\d\d\d\d\))", repl='')
movies['genres'] = movies.genres.str.replace('|', " ")

In [27]:
movies.head()

movieId                         title  \
0        1                    Toy Story    
1        2                      Jumanji    
2        3             Grumpier Old Men    
3        4            Waiting to Exhale    
4        5  Father of the Bride Part II    

                                        genres  Year  
0  Adventure Animation Children Comedy Fantasy  1995  
1                   Adventure Children Fantasy  1995  
2                               Comedy Romance  1995  
3                         Comedy Drama Romance  1995  
4                                       Comedy  1995

### Example on CountVectorizer

In [28]:
corpus = [
     'This is the first document.',                                  ### Line 1
    'This document is the second document.',                        ### Line 2
    'And this is the third one.',                                    ### Line 3
   'Is this the first document?',                                    ### Line 4
 ]

In [29]:
vectorizer = CountVectorizer()

In [30]:
X = vectorizer.fit_transform(corpus)

In [31]:
vectorizer.analyzer

'word'

In [32]:
vectorizer.get_feature_names

<bound method CountVectorizer.get_feature_names of CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)>

In [33]:
vectorizer.vocabulary_

{'this': 8,
 'is': 3,
 'the': 6,
 'first': 2,
 'document': 1,
 'second': 5,
 'and': 0,
 'third': 7,
 'one': 4}

In [34]:
print(X.toarray())                                             ### Row :- No. of line, Columns:- Words
                                                                        #### Draw Dence matrix of each Word 

[[0 1 1 1 0 0 1 0 1]
 [0 2 0 1 0 1 1 0 1]
 [1 0 0 1 1 0 1 1 1]
 [0 1 1 1 0 0 1 0 1]]


In [35]:
similarty_matrix = cosine_similarity(X, )

In [36]:
similarty_matrix                                      ### Cosine similarty with each line to another 
                                                     ### Row and Columns :- no. of Lines(4x4)

array([[1.        , 0.79056942, 0.54772256, 1.        ],
       [0.79056942, 1.        , 0.4330127 , 0.79056942],
       [0.54772256, 0.4330127 , 1.        , 0.54772256],
       [1.        , 0.79056942, 0.54772256, 1.        ]])

we have angular distance between line_1 to line_1 is 1 i.e (Cos theta = 0)
line_1 to line_2 have some angular distance.

## Let's get back to work

In [37]:
vector = CountVectorizer()

In [38]:
count_matrix = vector.fit_transform(movies['genres'])

In [39]:
count_matrix

<34208x24 sparse matrix of type '<class 'numpy.int64'>'
	with 71452 stored elements in Compressed Sparse Row format>

We have 24 types of geners with 34208 movies

### get Angular distance between these movies. to get this we will gonna be used Cosine_similarty that will gives us 34208 x34208 matrix

In [46]:
cosine_similarty_matrix = cosine_similarity(count_matrix[:5000])

In [48]:
similarty_df = pd.DataFrame(cosine_similarty_matrix, index=movies.movieId[:5000], columns=movies.movieId[:5000])

In [49]:
similarty_df.head()

movieId      1         2         3         4         5     6         7     \
movieId                                                                     
1        1.000000  0.774597  0.316228  0.258199  0.447214   0.0  0.316228   
2        0.774597  1.000000  0.000000  0.000000  0.000000   0.0  0.000000   
3        0.316228  0.000000  1.000000  0.816497  0.707107   0.0  1.000000   
4        0.258199  0.000000  0.816497  1.000000  0.577350   0.0  0.816497   
5        0.447214  0.000000  0.707107  0.577350  1.000000   0.0  0.707107   

movieId      8     9         10    ...      5085      5086      5087  \
movieId                            ...                                 
1        0.632456   0.0  0.258199  ...  0.000000  0.000000  0.258199   
2        0.816497   0.0  0.333333  ...  0.000000  0.000000  0.000000   
3        0.000000   0.0  0.000000  ...  0.000000  0.000000  0.816497   
4        0.000000   0.0  0.000000  ...  0.408248  0.408248  1.000000   
5        0.000000   0.0  0.000000  ...  0.000000  0.000000  0.577350   

movieId      5088      5089     5090      5091      5092  5093  5094  
movieId                                                               
1        0.258199  0.447214  0.00000  0.447214  0.632456   0.0   0.0  
2        0.000000  0.000000  0.00000  0.288675  0.408248   0.0   0.0  
3        0.408248  0.707107  0.00000  0.707107  0.500000   0.0   0.0  
4        0.666667  0.577350  0.57735  0.577350  0.408248   0.0   0.0  
5        0.577350  1.000000  0.00000  0.500000  0.707107   0.0   0.0  

[5 rows x 5000 columns]

In [43]:
rating =  pd.read_csv('D:\\Datasets\\Recommender_system\\ratings.csv')

In [52]:
user_viewed_movies = rating[rating.userId == 3]['movieId'].tolist()
user_viewed_movies

[356, 2394, 2431, 2445]

In [82]:
Top10_similar_movies=(similarty_df[user_viewed_movies].sum(axis = 1)).sort_values(ascending= False)[:10]

In [83]:
Top10_similar_movies

movieId
2583    2.414214
765     2.414214
4739    2.414214
4318    2.414214
3620    2.414214
1243    2.414214
1268    2.414214
1279    2.414214
178     2.414214
4759    2.414214
dtype: float64

In [99]:
for i in Top10_similar_movies.index:
    print(movies[movies.movieId==i]['title'].tolist()[0])

Cookie's Fortune 
Jack 
Lisa Picard is Famous (a.k.a. Famous) 
Postcards From the Edge 
Myth of Fingerprints, The 
Rosencrantz and Guildenstern Are Dead 
Pump Up the Volume 
Night on Earth 
Love & Human Remains 
Two Can Play That Game 
